### ISSUE TITLE
kubectl describe takes 60s to return

### ISSUE THREAD

**Issue by enisoc** [kubernetes/kubernetes#43648]
**Is this a request for help?**:

No.

**What keywords did you search in Kubernetes issues before filing this one?**:

* "kubectl describe"

---

**Is this a BUG REPORT or FEATURE REQUEST?**:

BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-rc.1", GitCommit:"8ea07d1fd277de8ab5ea7f281766760bcb7d0fe5", GitTreeState:"clean", BuildDate:"2017-03-24T19:19:36Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-rc.1", GitCommit:"8ea07d1fd277de8ab5ea7f281766760bcb7d0fe5", GitTreeState:"clean", BuildDate:"2017-03-24T19:07:50Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: GCE
- **OS** (e.g. from /etc/os-release):
  ```
  BUILD_ID=9000.84.2
  NAME="Container-Optimized OS"
  GOOGLE_CRASH_ID=Lakitu
  VERSION_ID=56
  BUG_REPORT_URL=https://crbug.com/new
  PRETTY_NAME="Container-Optimized OS from Google"
  VERSION=56
  GOOGLE_METRICS_PRODUCT_ID=26
  HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
  ID=cos
  ```
- **Kernel** (e.g. `uname -a`):
  ```
  Linux kubernetes-master 4.4.21+ #1 SMP Fri Feb 17 15:34:45 PST 2017 x86_64 Intel(R) Xeon(R) CPU @ 2.30GHz GenuineIntel GNU/Linux
  ```
- **Install tools**:
  ```
  cluster/kube-up.sh
  cluster/gce/upgrade.sh
  ```
- **Others**:

**What happened**:

Running `kubectl describe` on anything takes almost exactly 60s to return a result every time. It does eventually return a result that looks correct.

**What you expected to happen**:

The result should be returned in less than 60s.

**How to reproduce it** (as minimally and precisely as possible):

I encountered this after an upgrade, but I don't know if that's actually required to reproduce it.

The key seems to be generating a lot of events first. The way I ended up with lots of events was by intentionally tickling a bug in 1.5 before upgrading, by creating all the objects in [overlap.yaml](https://gist.github.com/enisoc/4f134a94359719e0f39bbd54f8711a90) at the same timestamp. After upgrading to 1.6, the controllers stop fighting but the events remain.

```
# Start a v1.5.4 cluster.
cd v1.5.4
cluster/get-kube-binaries.sh
cluster/kube-up.sh

# Create overlapping controllers that fight and generate lots of events.
kubectl create -f overlap.yaml

# Wait a few minutes to make sure it generates enough events.
# Watch this while you wait...
kubectl get po -w

# Upgrade the master to v1.6.0-rc.1
cd ../v1.6.0-rc.1
cluster/get-kube-binaries.sh
cluster/gce/upgrade.sh -M v1.6.0-rc.1

# Create something.
kubectl run nginx --image=nginx

# Try to describe it, and note how long it takes.
time kubectl describe deploy nginx
```

**Anything else we need to know**:

The problem is the same if I use kubectl v1.5.4:

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.4", GitCommit:"7243c69eb523aa4377bce883e7c0dd76b84709a1", GitTreeState:"clean", BuildDate:"2017-03-07T23:53:09Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-rc.1", GitCommit:"8ea07d1fd277de8ab5ea7f281766760bcb7d0fe5", GitTreeState:"clean", BuildDate:"2017-03-24T19:07:50Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

If I do `kubectl get -o yaml` instead, it behaves normally.

**Comment by liggitt** [comment]
If you do this with `--v=6`, which call(s) are taking the time?

**Comment by ixdy** [comment]
I just deployed a v1.6.0-rc.1 cluster on gce using kube-up.sh, and it's working as normal (no 60s-delays).

**Comment by enisoc** [comment]
I retried my own repro steps from scratch and now I can't repro. So I must have done something else to get into a bad state. I will look at other possible repro steps that I may have missed.

**Comment by enisoc** [comment]
I was able to repro again. The key seems to be generating a large amount of events first.

Here is the output from `kubectl describe deploy asd --v=6`:

```
I0324 16:34:11.448308   31776 round_trippers.go:417] GET https://35.184.16.161/apis/extensions/v1beta1/namespaces/default/deployments/asd 200 OK in 221 milliseconds
I0324 16:34:11.491513   31776 round_trippers.go:417] GET https://35.184.16.161/apis/extensions/v1beta1/namespaces/default/deployments/asd 200 OK in 39 milliseconds
I0324 16:34:11.532784   31776 round_trippers.go:417] GET https://35.184.16.161/apis/extensions/v1beta1/namespaces/default/replicasets?labelSelector=run%3Dasd 200 OK in 40 milliseconds
I0324 16:34:11.575619   31776 round_trippers.go:417] GET https://35.184.16.161/api/v1/namespaces/default/pods?labelSelector=run%3Dasd 200 OK in 42 milliseconds
I0324 16:35:11.614114   31776 round_trippers.go:417] GET https://35.184.16.161/api/v1/namespaces/default/events?fieldSelector=involvedObject.kind%3DDeployment%2CinvolvedObject.name%3Dasd%2CinvolvedObject.namespace%3Ddefault%2CinvolvedObject.uid%3D3ff157aa-10ea-11e7-afa6-42010a800002 504 Gateway Timeout in 60037 milliseconds
```

In my case, the way I ended up with a large number of events was by tickling a bug in 1.5 before doing the upgrade. I have a battery of overlapping controllers that causes a 1.5 cluster to freak out if you create them all at once:

https://gist.github.com/enisoc/4f134a94359719e0f39bbd54f8711a90

**Comment by liggitt** [comment]
Yeah, with sufficiently large numbers of objects, you can hit LB or proxy timeouts. Non-collapsing events (unique messages that include a timestamp or resourceVersion for different events for the same cause and object) for misbehaving components are a likely cause of that.

**Comment by enisoc** [comment]
Is this a pre-existing known issue? If so, I will remove it from the milestone.

**Comment by liggitt** [comment]
Issues with enormous event volume is known: https://github.com/kubernetes/kubernetes/issues/4432

Specifically about event compression not interacting well with fighting controllers: https://github.com/kubernetes/kubernetes/pull/4372/files#r24559200

**Comment by enisoc** [comment]
I see. In that case, I will close this since the only way I know of to cause it is fighting controllers, which is already fixed in 1.6.

**Comment by liggitt** [comment]
The TTL on event values in etcd is supposed to backstop it a little by clearing out the flood eventually, though it doesn't help at all while you have the flood
